Find a functionally equivalent code:android.os.Parcel.readArrayMapInternal:COMMENT
Method Modifier: 
Comment:/* package */

@@ -4,15 +4,19 @@
         here.fillInStackTrace();
         Log.d(TAG, "Reading " + N + " ArrayMap entries", here);
     }
+    int startPos;
     while (N > 0) {
-        Object key = readValue(loader);
         if (DEBUG_ARRAY_MAP)
-            Log.d(TAG, "  Read #" + (N - 1) + ": key=0x" + (key != null ? key.hashCode() : 0) + " " + key);
+            startPos = dataPosition();
+        String key = readString();
         Object value = readValue(loader);
+        if (DEBUG_ARRAY_MAP)
+            Log.d(TAG, "  Read #" + (N - 1) + " " + (dataPosition() - startPos) + " bytes: key=0x" + Integer.toHexString((key != null ? key.hashCode() : 0)) + " " + key);
         outVal.append(key, value);
         N--;
     }
+    outVal.validate();
 }

